src/lightgbm/src/main/scala/LightGBMUtils.scala

```diff
@@ -0,0 +1,216 @@
+    lightgbmlib.longp_assign(longPtr, value)
+    lightgbmlib.long_to_int64_t_ptr(longPtr)
+  }
+
+  def generateData(numRows: Int, rowsAsDoubleArray: Array[Array[Double]]): (SWIGTYPE_p_void, SWIGTYPE_p_double) = {
+    val numCols = rowsAsDoubleArray.head.length
+    val data = lightgbmlib.new_doubleArray(numCols * numRows)
+    rowsAsDoubleArray.zipWithIndex.foreach(ri =>
+      ri._1.zipWithIndex.foreach(value =>
+        lightgbmlib.doubleArray_setitem(data, value._2 + (ri._2 * numCols), value._1)))
+    (lightgbmlib.double_to_voidp_ptr(data), data)
+  }
+
+  def generateDenseDataset(numRows: Int, rowsAsDoubleArray: Array[Array[Double]]): SWIGTYPE_p_void = {
+    val numRowsIntPtr = lightgbmlib.new_intp()
+    lightgbmlib.intp_assign(numRowsIntPtr, numRows)
+    val numRows_int32_tPtr = lightgbmlib.int_to_int32_t_ptr(numRowsIntPtr)
+    val numCols = rowsAsDoubleArray.head.length
+    val isRowMajor = 1
+    val numColsIntPtr = lightgbmlib.new_intp()
+    lightgbmlib.intp_assign(numColsIntPtr, numCols)
+    val numCols_int32_tPtr = lightgbmlib.int_to_int32_t_ptr(numColsIntPtr)
+    val datasetOutPtr = lightgbmlib.voidpp_handle()
+    val datasetParams = "max_bin=255 is_pre_partition=True"
+    val data64bitType = lightgbmlibConstants.C_API_DTYPE_FLOAT64
+    val data = generateData(numRows, rowsAsDoubleArray)
+
+    // Generate the dataset for features
+    LightGBMUtils.validate(lightgbmlib.LGBM_DatasetCreateFromMat(data._1, data64bitType,
+      numRows_int32_tPtr, numCols_int32_tPtr, isRowMajor, datasetParams, null, datasetOutPtr), "Dataset create")
+    lightgbmlib.delete_doubleArray(data._2)
+    lightgbmlib.voidpp_value(datasetOutPtr)
+  }
+
+  /** Generates a sparse dataset in CSR format.
+    * @param sparseRows The rows of sparse vector.
+    * @return
+    */
+  def generateSparseDataset(sparseRows: Array[SparseVector]): SWIGTYPE_p_void = {
+    val valuesArray = sparseRows.flatMap(_.values)
+    val values = newDoubleArray(valuesArray)
+    val indexesArray = sparseRows.flatMap(_.indices)
+    val indexes = newIntArray(indexesArray)
+    val indptr = new Array[Int](sparseRows.length + 1)
+    sparseRows.zipWithIndex.foreach {
+      case (row, index) => indptr(index + 1) = indptr(index) + row.numNonzeros
+    }
+    val indptrNative = newIntArray(indptr)
+    val numCols = sparseRows(0).size
+
+    val datasetOutPtr = lightgbmlib.voidpp_handle()
+    val datasetParams = "max_bin=255 is_pre_partition=True"
+    val dataInt32bitType = lightgbmlibConstants.C_API_DTYPE_INT32
+    val data64bitType = lightgbmlibConstants.C_API_DTYPE_FLOAT64
+
+    // Generate the dataset for features
+    LightGBMUtils.validate(CSRUtils.LGBM_DatasetCreateFromCSR(indptrNative._1, dataInt32bitType,
+      indexes._1, values._1, data64bitType, intToPtr(indptr.length), intToPtr(valuesArray.length), intToPtr(numCols),
+      datasetParams, null, datasetOutPtr), "Dataset create")
+    // Delete the input rows
+    lightgbmlib.delete_doubleArray(values._2)
```

Comment:
deletes must always be inside try..finally otherwise you will leak memory in the presence of exceptions. Could you please verify other places where you have delete_foo or delete_bar and ensure they are inside a try..finally guard?